### PR TITLE
[15.0] [ADD] base_address_parish

### PR DESCRIPTION
**Pull request by mamfredym**
Create a new module for issue 

https://github.com/OCA/partner-contact/issues/1214

### CODE REVIEW

File: base_address_parish/models/res_country.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+from odoo import fields, models
+
+
+class Country(models.Model):
+    _inherit = "res.country"
+
+    enforce_parishes = fields.Boolean(
+        help="Check this box to ensure every address created in that country has a "
+        "Parish' chosen in the list of the city's parishes.",
+    )
```
Comment (celm1990) [inline base_address_parish/models/res_country.py:13]: what is use for this field???
Field `parish_id  `is ever show on partner view, i think than attrs must be based on this field

File: base_address_parish/models/res_parish.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    state_id = fields.Many2one(
+        "res.country.state",
+        "State",
+        readonly=True,
+        domain="[('country_id', '=', country_id)]",
+        related="city_id.state_id",
+    )
+    country_id = fields.Many2one(
+        "res.country", string="Country", readonly=True, related="city_id.country_id"
+    )
```
Comment (celm1990) [inline base_address_parish/models/res_parish.py]: Fields related can be `store=True`? for can grouping/search if need

File: base_address_parish/models/res_parish.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    _order = "name"
+
+    name = fields.Char(required=True, translate=True)
+    zip = fields.Char()
+    city_id = fields.Many2one("res.city", "City", required=True)
+    state_id = fields.Many2one(
+        "res.country.state",
+        "State",
+        readonly=True,
+        domain="[('country_id', '=', country_id)]",
```
Comment (celm1990) [inline base_address_parish/models/res_parish.py]: `domain `is unnecesary if field is related and readonly=True

File: base_address_parish/views/res_parish_view.xml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        </record>
+
+        <!--MENU ITEMS FOR CITIES-->
+        <menuitem
+        name="Cities"
+        id="locations_menu_cities"
+        parent="contacts.menu_localisation"
+        action="base_address_city.action_res_city_tree"
+        sequence="4"
+    />
```
Comment (celm1990) [inline base_address_parish/views/res_parish_view.xml]: I see menu is added on other module([base_location ](https://github.com/OCA/partner-contact/blob/15.0/base_location/views/res_city_view.xml#L53)) and see two menus. I think this module should not add menu

image

**Review by celm1990 — CHANGES_REQUESTED** [review]

File: base_address_parish/models/res_country.py (hunk 1 above)
Comment (mamfredym) [reply base_address_parish/models/res_country.py:13]: The parish concept is based for country. Each country define if have or not parishes

File: base_address_parish/models/res_parish.py (hunk 2 above)
Comment (mamfredym) [reply base_address_parish/models/res_parish.py]: Sure

File: base_address_parish/models/res_parish.py (hunk 3 above)
Comment (mamfredym) [reply base_address_parish/models/res_parish.py]: OK

File: base_address_parish/views/res_parish_view.xml (hunk 4 above)
Comment (mamfredym) [reply base_address_parish/views/res_parish_view.xml]: Solved

File: base_address_parish/models/res_country.py (hunk 1 above)
Comment (celm1990) [reply base_address_parish/models/res_country.py:13]: my point is: field `enforce_parishes  `has not use on any view on `attrs` or python code

File: base_address_parish/views/res_partner_view.xml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      <field name="model">res.partner</field>
+      <field name="inherit_id" ref="base.view_partner_address_form" />
+      <field name="arch" type="xml">
+        <field name="street2" position="after">
+          <field
+                    name="parish_id"
+                    placeholder="Parish..."
+                    class="o_address_street"
+                    options="{'no_open': True, 'no_quick_create': True}"
+                    attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)]}"
```
Comment (celm1990) [inline base_address_parish/views/res_partner_view.xml]:
```suggestion
                    attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)], 'invisible': [('enforce_parishes', '=', False)]}"
```

File: base_address_parish/views/res_partner_view.xml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      <field name="model">res.partner</field>
+      <field name="inherit_id" ref="base.res_partner_view_form_private" />
+      <field name="arch" type="xml">
+          <field name="street2" position="after">
+            <field
+                    name="parish_id"
+                    placeholder="Parish..."
+                    class="o_address_street"
+                    options="{'no_open': True, 'no_quick_create': True}"
+                    attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)]}"
```
Comment (celm1990) [inline base_address_parish/views/res_partner_view.xml]:
```suggestion
                    attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)], 'invisible': [('enforce_parishes', '=', False)]}"
```

File: base_address_parish/views/res_partner_view.xml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      <field name="model">res.partner</field>
+      <field name="inherit_id" ref="base.view_partner_form" />
+      <field name="arch" type="xml">
+        <field name="street2" position="after">
+          <field
+                    name="parish_id"
+                    placeholder="Parish..."
+                    class="o_address_street"
+                    options="{'no_open': True, 'no_quick_create': True}"
+                    attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)]}"
```
Comment (celm1990) [inline base_address_parish/views/res_partner_view.xml]:
```suggestion
                    attrs="{'readonly': [('type', '=', 'contact'),('parent_id', '!=', False)], 'invisible': [('enforce_parishes', '=', False)]}"
```

**Review by celm1990 — CHANGES_REQUESTED** [review]

File: base_address_parish/models/res_country.py (hunk 1 above)
Comment (cormaza) [reply base_address_parish/models/res_country.py:13]: I think that this field if you want keep it, should be implemented like

https://github.com/odoo/odoo/blob/a19d81bcab978150cf01f1ae85543adc5cb92e5e/addons/base_address_city/models/res_partner.py#L56

File: base_address_parish/views/res_partner_view.xml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      <field name="name">parish.partner.view</field>
+      <field name="model">res.partner</field>
+      <field name="inherit_id" ref="base.view_partner_form" />
+      <field name="arch" type="xml">
+        <field name="street2" position="after">
+          <field
+                    name="parish_id"
+                    placeholder="Parish..."
+                    class="o_address_street"
+                    options="{'no_open': True, 'no_quick_create': True}"
```
Comment (celm1990) [inline base_address_parish/views/res_partner_view.xml]: Parish Records are static only if changed territorial organization of country is need create a new record. So, i think users can not create new records here, it must be do in apropied menu
```suggestion
                    options="{'no_open': True, 'no_create': True}"
```

File: base_address_parish/views/res_partner_view.xml [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      <field name="name">parish.res.partner.form.address</field>
+      <field name="model">res.partner</field>
+      <field name="inherit_id" ref="base.view_partner_address_form" />
+      <field name="arch" type="xml">
+        <field name="street2" position="after">
+          <field
+                    name="parish_id"
+                    placeholder="Parish..."
+                    class="o_address_street"
+                    options="{'no_open': True, 'no_quick_create': True}"
```
Comment (celm1990) [inline base_address_parish/views/res_partner_view.xml]: Same here
```suggestion
                    options="{'no_open': True, 'no_create': True}"
```

File: base_address_parish/views/res_partner_view.xml [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+      <field name="name">parish.res.partner.form.private</field>
+      <field name="model">res.partner</field>
+      <field name="inherit_id" ref="base.res_partner_view_form_private" />
+      <field name="arch" type="xml">
+          <field name="street2" position="after">
+            <field
+                    name="parish_id"
+                    placeholder="Parish..."
+                    class="o_address_street"
+                    options="{'no_open': True, 'no_quick_create': True}"
```
Comment (celm1990) [inline base_address_parish/views/res_partner_view.xml]: and here
```suggestion
                    options="{'no_open': True, 'no_create': True}"
```

**Review by celm1990 — CHANGES_REQUESTED** [review]

File: base_address_parish/models/res_country.py (hunk 1 above)
Comment (mamfredym) [reply base_address_parish/models/res_country.py:13]: The suggestion has added in the last commit today

File: base_address_parish/models/res_partner.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    @api.onchange("parish_id")
+    def _onchange_parish_id(self):
+        if self.parish_id:
+            self.zip = self.parish_id.zip
+            self.city_id = self.parish_id.city_id
+            self.state_id = self.parish_id.state_id
+        elif self._origin:
+            self.city_id = False
+            self.zip = False
+            self.state_id = False
```
Comment (celm1990) [inline base_address_parish/models/res_partner.py]:
```suggestion
          self.zip = self.parish_id.zip or ""
          self.city_id = self.parish_id.city_id or False
          self.state_id = self.parish_id.state_id or False
```

**Review by celm1990 — CHANGES_REQUESTED** [review]

File: base_address_parish/models/res_partner.py (hunk 11 above)
Comment (mamfredym) [reply base_address_parish/models/res_partner.py]: Perfect

File: base_address_parish/tests/test_res_partner.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    @classmethod
+    def setUpClass(cls):
+        super(TestResPartner, cls).setUpClass()
+
+        # models
+        cls.country_model = cls.env["res.country"]
+        cls.states_model = cls.env["res.country.state"]
+        cls.cities_model = cls.env["res.city"]
+
+        cls.country = cls.country_model.search([("id", "=", cls.env.ref("base.ec").id)])
```
Comment (celm1990) [inline base_address_parish/tests/test_res_partner.py]:
```suggestion
        cls.country = cls.env.ref("base.ec")
```

File: base_address_parish/tests/test_res_partner.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
+# Copyright 2022: PBox (<https://www.pupilabox.net.ec>)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo.tests import common
+from odoo.tests.common import Form
+
+
+class TestResPartner(common.TransactionCase):
+    def setUp(self):
+        super(TestResPartner, self).setUp()
```
Comment (celm1990) [inline base_address_parish/tests/test_res_partner.py:10]: ????

File: base_address_parish/tests/test_res_partner.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+        cls.country_model = cls.env["res.country"]
+        cls.states_model = cls.env["res.country.state"]
+        cls.cities_model = cls.env["res.city"]
+
+        cls.country = cls.country_model.search([("id", "=", cls.env.ref("base.ec").id)])
+        cls.country.write({"enforce_parishes": True, "enforce_cities": True})
+
+        cls.states = cls.env["res.country.state"].search(
+            [("country_id", "=", cls.env.ref("base.ec").id)]  # In Ecuador uses Parishes
+        )
```
Comment (celm1990) [inline base_address_parish/tests/test_res_partner.py]:
```suggestion
        cls.states = cls.env["res.country.state"].search(
            [("country_id", "=", cls.country.id)]  # In Ecuador uses Parishes
        )
```

File: base_address_parish/tests/test_res_partner.py (hunk 13 above)
Comment (mamfredym) [reply base_address_parish/tests/test_res_partner.py:10]: In anothers modules call super

File: base_address_parish/tests/test_res_partner.py (hunk 12 above)
Comment (mamfredym) [reply base_address_parish/tests/test_res_partner.py]: Perfect, thanks

File: base_address_parish/tests/test_res_partner.py (hunk 14 above)
Comment (mamfredym) [reply base_address_parish/tests/test_res_partner.py]: Perfect, thanks

**Comment by celm1990** [comment]
@cormaza  @mamfredym 
Please rebase commits to head for correct pre-commit, this PR is need for https://github.com/OCA/l10n-ecuador/pull/36

**Comment by celm1990** [comment]
@mamfredym
Could you please complete this PR or let me know if it needs to be suspended and work on another PR?

@cormaza, your comments are blocking progress on this PR: https://github.com/OCA/l10n-ecuador/pull/39. Please review your comments and, if possible, provide suggestions or solutions to move forward.

**Comment by cormaza** [comment]
> @mamfredym Could you please complete this PR or let me know if it needs to be suspended and work on another PR?
> 
> @cormaza, your comments are blocking progress on this PR: [OCA/l10n-ecuador#39](https://github.com/OCA/l10n-ecuador/pull/39). Please review your comments and, if possible, provide suggestions or solutions to move forward.

I'm not sure what comments you are asking, but in pr you refer I think is better use class created by this module, you should use "res.parish" instead of "l10n.ec.parish"

File: base_address_parish/tests/test_res_partner.py (hunk 13 above)
Comment (NL66278) [inline base_address_parish/tests/test_res_partner.py:10]: Just use ` super()`

File: base_address_parish/tests/test_res_partner.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+from odoo.tests.common import Form
+
+
+class TestResPartner(common.TransactionCase):
+    def setUp(self):
+        super(TestResPartner, self).setUp()
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestResPartner, cls).setUpClass()
```
Comment (NL66278) [inline base_address_parish/tests/test_res_partner.py:14]: Just ` super()`  here as well.

**Pull request closed without merging**